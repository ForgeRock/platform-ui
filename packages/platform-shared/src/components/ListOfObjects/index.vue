<!-- Copyright (c) 2020-2021 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div class="pt-2">
    <div :class="{ 'border-bottom': isValidJSONString(listValues) && isValidField()}">
      <div class="d-flex justify-content-between align-items-center">
        <label>{{ capitalizedDescription }}</label>
      </div>
      <div>
        <div
          v-if="!listValues || !listValues.length"
          class="d-flex pt-3 pb-3 px-0 border-top align-items-center">
          <div class="text-muted text-left flex-grow-1">
            ({{ $t('common.none') }})
          </div>
          <button
            class="btn btn-outline-secondary mr-1 mb-2 mb-lg-0"
            @click.prevent="addObjectToList(-1)">
            <FrIcon
              name="add"
            />
          </button>
        </div>
        <template v-if="isValidJSONString(listValues) && isValidField()">
          <div
            v-for="(obj, index) in listValues"
            :key="obj.listUniqueIndex"
            class="d-flex pt-3 pb-2 px-0 border-top">
            <div class="flex-grow-1 pr-3 position-relative">
              <div class="form-row align-items-center">
                <div
                  v-for="(objValue, key) in obj"
                  :key="key"
                  class="col-lg-4 pb-2">
                  <div
                    v-if="key !== 'listUniqueIndex'"
                    class="position-relative">
                    <div v-if="properties[key].type === 'boolean'">
                      <BFormCheckbox
                        v-model="obj[key]"
                        :name="key">
                        {{ key }}
                      </BFormCheckbox>
                    </div>
                    <div v-else-if="properties[key].type === 'number'">
                      <FrField
                        v-model.number="obj[key]"
                        type="number"
                        validation="required|numeric"
                        :label="key"
                        @input="emitInput(listValues)"
                      />
                    </div>
                    <div v-else>
                      <FrField
                        v-model="obj[key]"
                        :label="key"
                        :type="properties[key].type"
                        :validation="required && required.length && required.includes(properties[key].title) ? 'required' : ''"
                        @input="emitInput(listValues)"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="position-relative d-inline-flex justify-content-end"
                style="width: 128px;">
                <button
                  class="btn btn-outline-secondary mr-1 mb-2 mb-lg-0"
                  @click.prevent="removeElementFromList(index)">
                  <FrIcon
                    name="remove"
                  />
                </button>
                <button
                  class="btn btn-outline-secondary mr-1 mb-2 mb-lg-0"
                  @click.prevent="addObjectToList(index)">
                  <FrIcon
                    name="add"
                  />
                </button>
              </div>
            </div>
          </div>
        </template>
        <div v-else>
          <FrInlineJsonEditor
            v-on="$listeners"
            language="json"
            :line-count="lineCount"
            :read-only="false"
            :value="advancedValue"
            @update-field="$emit('input', $event)" />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import {
  cloneDeep,
} from 'lodash';
import {
  BFormCheckbox,
} from 'bootstrap-vue';
import FrField from '@forgerock/platform-shared/src/components/Field';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import FrInlineJsonEditor from '@forgerock/platform-shared/src/components/InlineJsonEditor';
import ListsMixin from '@forgerock/platform-shared/src/mixins/ListsMixin';

/**
 * @description Component that provides support for list of objects
 *
 * Attempts to render list of obects. If the object contains complex properties (i.e. nested arrays or objects)
 * or is the array is not valid JSON, displays array of objects in code editor
 */
export default {
  name: 'ListOfObjects',
  components: {
    BFormCheckbox,
    FrField,
    FrIcon,
    FrInlineJsonEditor,
  },
  mixins: [
    ListsMixin,
  ],
  props: {
    description: {
      type: String,
      default: '',
    },
    properties: {
      type: Object,
      default: () => {},
    },
    label: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      default: '',
    },
    required: {
      type: Array,
      default: () => [],
    },
    value: {
      type: [Array, String],
      default: () => [],
    },
  },
  data() {
    return {
      expanded: false,
      hover: false,
      isValidJason: true,
      listValues: [],
      // eslint-disable-next-line vue/no-reserved-keys
      listUniqueIndex: 0,
    };
  },
  computed: {
    advancedValue() {
      return this.listValues.map((val) => {
        delete val.listUniqueIndex;
        return val;
      });
    },
  },
  mounted() {
    if (this.value && this.value !== '' && this.value !== null) {
      const listValues = cloneDeep(this.value);
      listValues.forEach((val) => {
        val.listUniqueIndex = this.getUniqueIndex();
      });
      this.listValues = listValues;
    }
  },
  methods: {
    /**
     * populate list of objects with new member.  Set defaults for boolean and number properties
     */
    addObjectToList(valueIndex) {
      const emptyObjectWithKeys = this.createObject(this.properties);
      this.listValues.splice(valueIndex + 1, 0, { ...emptyObjectWithKeys, listUniqueIndex: this.getUniqueIndex() });
      this.emitInput(this.listValues);
    },
    emitInput(value) {
      const emitValue = cloneDeep(value);
      emitValue.map((val) => {
        delete val.listUniqueIndex;
        return val;
      });
      if (emitValue.length === 0) {
        this.$emit('input', '');
      } else {
        this.$emit('input', emitValue);
      }
    },
    /**
     * Ensures our keys in v-if iteration have unique values
     *
     * @returns {number} New unique index
     */
    getUniqueIndex() {
      this.listUniqueIndex += 1;
      return this.listUniqueIndex;
    },
    /**
     * determine whether any properties of an object in the array
     * are too complex to render (i.e. objects with sub object or array properties)
     */
    isValidField() {
      const values = Object.values(this.properties) || [];
      const results = [];

      values.forEach((val) => {
        if (val.type && (val.type === 'array' || val.type === 'object')) {
          results.push(false);
        }
        results.push(true);
      });

      return !results.includes(false);
    },
    /**
     * remove element from list component at index
     */
    removeElementFromList(index) {
      this.listValues.splice(index, 1);
      this.emitInput(this.listValues);
    },
  },
};
</script>
