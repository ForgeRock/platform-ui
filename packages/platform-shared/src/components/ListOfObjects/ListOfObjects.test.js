/**
 * Copyright (c) 2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { shallowMount } from '@vue/test-utils';
import ListOfObjects from './index';

describe('ListOfObjects', () => {
  it('ListOfObjects loaded', () => {
    const wrapper = shallowMount(ListOfObjects, {
      mocks: {
        $t: () => {},
      },
      propsData: {
        items: {},
        properties: {},
      },
    });
    expect(wrapper.name()).toBe('ListOfObjects');
  });

  it('ListOfObjects sets listValues when there is a value', () => {
    const wrapper = shallowMount(ListOfObjects, {
      mocks: {
        $t: () => {},
      },
      propsData: {
        items: {},
        properties: {
          testPropertyKey: {
            type: 'array',
            value: 'testPropertyValue',
          },
        },
        name: 'test',
        value: [{ value: 'test' }],
      },
    });
    expect(wrapper.vm.listValues).toStrictEqual([{
      value: 'test',
    }]);
  });

  it('ListOfObjects adds elements to list', () => {
    const wrapper = shallowMount(ListOfObjects, {
      mocks: {
        $t: () => {},
      },
      propsData: {
        items: {},
        properties: {
          testPropertyKey: {
            type: 'array',
            value: 'testPropertyValue',
            items: {
              type: 'other',
            },
          },
        },
        name: 'test',
        value: [{ value: 'test' }],
      },
    });
    wrapper.vm.addObjectToList(0);
    expect(wrapper.vm.listValues).toStrictEqual([{
      value: 'test',
    }, {
      testPropertyKey: [],
    }]);
    wrapper.setProps({
      properties: {
        testPropertyKey: {
          type: 'array',
          value: 'testPropertyValue',
          items: {
            type: 'array',
            value: 'testPropertyValue',
            items: {
              type: 'other',
            },
          },
        },
      },
    });
    wrapper.vm.addObjectToList(1);
    expect(wrapper.vm.listValues).toStrictEqual([{
      value: 'test',
    }, {
      testPropertyKey: [],
    }, {
      testPropertyKey: [[]],
    }]);
  });

  it('ListOfObjects removes elements from list', () => {
    const wrapper = shallowMount(ListOfObjects, {
      mocks: {
        $t: () => {},
      },
      propsData: {
        items: {},
        properties: {
          testPropertyKey: {
            type: 'array',
            value: 'testPropertyValue',
            items: {
              type: 'other',
            },
          },
        },
        name: 'test',
        value: [{ value: 'test' }],
      },
    });
    wrapper.vm.addObjectToList(0);
    expect(wrapper.vm.listValues).toStrictEqual([{
      value: 'test',
    }, {
      testPropertyKey: [],
    }]);

    wrapper.vm.removeElementFromList(1);
    expect(wrapper.vm.listValues).toStrictEqual([{
      value: 'test',
    }]);
  });

  it('ListOfObjects checks if field is valid', () => {
    const wrapper = shallowMount(ListOfObjects, {
      mocks: {
        $t: () => {},
      },
      propsData: {
        items: {},
        properties: {
          testPropertyKey: {
            type: 'array',
            value: [
              {
                testPropertyKey: 'testPropertyValue',
              },
            ],
            items: {
              type: 'other',
            },
          },
        },
        name: 'test',
        value: [{ testPropertyKey: 'test' }],
      },
    });
    expect(wrapper.vm.isValidField()).toBe(false);
    wrapper.setProps({
      properties: {
        testPropertyKey: {
          type: 'object',
          value: [
            {
              testPropertyKey: 'testPropertyValue',
            },
          ],
          items: {
            type: 'other',
          },
        },
      },
    });
    expect(wrapper.vm.isValidField()).toBe(false);
    wrapper.setProps({
      properties: {
        testPropertyKey: {
          type: 'other',
          value: [
            {
              testPropertyKey: 'testPropertyValue',
            },
          ],
          items: {
            type: 'other',
          },
        },
      },
    });
    expect(wrapper.vm.isValidField()).toBe(true);
  });
});
