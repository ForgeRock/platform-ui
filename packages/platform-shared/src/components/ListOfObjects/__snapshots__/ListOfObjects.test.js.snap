// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListOfObjects Add button disabled if field is disabled: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects Generates an uuid as a name for the useField method: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "value=",
    [
      {
        "test": "test",
      },
    ],
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Property "name" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "value=",
    [
      {
        "test": "test",
      },
    ],
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects If all fields are empty or null, remove the object: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "label="test"",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testtwo": {
        "title": "testtwo",
        "type": "string",
      },
    },
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects ListOfObjects adds empty elements to list: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_V_MODEL) v-model usage on component has changed in Vue 3. Component that expects to work with v-model should now use the "modelValue" prop and emit the "update:modelValue" event. You can update the usage and then opt-in to Vue 3 behavior on a per-component basis with \`compatConfig: { COMPONENT_V_MODEL: false }\`.
  Details: https://v3-migration.vuejs.org/breaking-changes/v-model.html",
    "
",
    " at <BFormCheckbox",
    "modelValue=false",
    "onUpdate:modelValue=fn<onUpdateModelValue>",
    "disabled=undefined",
    " ...",
    ">",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "checked" with v-bind value \`false\` will render checked="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "autofocus" with v-bind value \`false\` will render autofocus="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "disabled" with v-bind value \`false\` will render disabled="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "required" with v-bind value \`false\` will render required="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "plain" with v-bind value \`false\` will render plain="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "button" with v-bind value \`false\` will render button="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "inline" with v-bind value \`false\` will render inline="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "indeterminate" with v-bind value \`false\` will render indeterminate="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "switch" with v-bind value \`false\` will render switch="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "uncheckedValue" with v-bind value \`false\` will render uncheckedValue="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormCheckbox",
    "disabled=undefined",
    "name="testPropertyKey_0_16"",
    "onChange=fn",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects ListOfObjects checks if field is valid: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "testPropertyKey": {
        "title": "testPropertyKey",
        "type": "array",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "testPropertyKey": {
        "title": "testPropertyKey",
        "type": "array",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "readOnly" with v-bind value \`false\` will render readOnly="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <FrInlineJsonEditor",
    "language="json"",
    "line-count=1",
    "read-only=false",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "testPropertyKey": {
        "title": "testPropertyKey",
        "type": "array",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects ListOfObjects removes elements from list: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects ListOfObjects sets listValues and converts to array when there is an object value: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "multiValued=false",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`ListOfObjects ListOfObjects sets listValues when there is an array value: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) (2)",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "value=",
    [
      {
        "test": "test",
      },
    ],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: Property "description" was accessed during render but is not defined on instance.",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "value=",
    [
      {
        "test": "test",
      },
    ],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "canEnterPlaceholders" with v-bind value \`false\` will render canEnterPlaceholders="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <FrField",
    "value="test"",
    "onInput=fn<onInput>",
    "disabled=undefined",
    " ...",
    ">",
    "
",
    " at <ListOfObjects",
    "properties=",
    {
      "test": {
        "title": "test",
        "type": "string",
      },
      "testPropertyKey": {
        "title": "testPropertyValue",
        "type": "boolean",
      },
    },
    "label="test"",
    "value=",
    [
      {
        "test": "test",
      },
    ],
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
